### DISCUSSION TITLE
Use SIDs from local rules in other local rules

### DISCUSSION THREAD

**Discussion by martinzeifang** [wazuh/wazuh#15320] [Q&A] [answered]
Hi everyone, 

I have a question: 
I have a local rule file. let's call it Custom01.xml with a rule id. For example: 

```
 <rule id="100001" level="5">
    <if_sid>5716</if_sid>
    <srcip>1.1.1.1</srcip>
    <description>sshd: authentication failed from IP 1.1.1.1.</description>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
  </rule>
```

and a second rule file, let's call it Custom02.xml with a rule using the id from custom01.xml: 

```
 <rule id="200001" level="12">
    <if_sid>100001</if_sid>
    <match>superadmin</match>
    <description>sshd: authentication failed from IP 1.1.1.1. as superadmin</description>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
  </rule>
```

in Wazuh 4.2 I couldn't save the second file with an error message. Since 4.3 I can save the file but the rule never generates an alert. 
is it supported to use ids from local rule files in other files? I can't find anything in the documentation about it. 

thanks everyone

**Comment by f-galland** [comment, marked as answer]
Hi @martinzeifang,

As a first consideration, remember to keep your custom rule IDs between 100000 and 120000.
https://documentation.wazuh.com/current/user-manual/ruleset/custom.html#adding-new-decoders-and-rules.



In order to pin the reason your rule isn't getting triggered out, we would need to have an example log line from ssh for your desired use-case.

However, there are a couple of things we can improve:

Your first rule is actually fully muting rule 5716, since it's only the last rule in the chain that generates an alert. In order to avoid that, you would need to make rule 100001 a sibling of 5716, not a child.

If you go through the rule file for ssh:
https://github.com/wazuh/wazuh/blob/46dcb5799ed5c3695d8bfde90cea011b7cf06170/ruleset/rules/0095-sshd_rules.xml#L149

You will see that rule 5716 is a child of 5700. That's the one we need to make our top level rule.
In order to match the same criteria of 5716, plus the intended srcip match, you can just add that to the rule itself.

The same thing applies to the second rule.

One last consideration is that your rule blocks need to be enclosed in a <group> tag as well.

So a more proper custom_ssh_rules.xml file would look like this:



```
root@manager:~# cat /var/ossec/etc/rules/custom_ssh_rules.xml
<group name="syslog,sshd,">

  <rule id="110050" level="5">
    <if_sid>5700</if_sid>
    <match>^Failed|^error: PAM: Authentication</match>
    <srcip>1.1.1.1</srcip>
    <description>sshd: authentication failed from IP: $(srcip)</description>
    <mitre>
      <id>T1110</id>
    </mitre>
    <group>authentication_failed,gdpr_IV_35.7.d,gdpr_IV_32.2,gpg13_7.1,hipaa_164.312.b,nist_800_53_AU.14,nist_800_53_AC.7,pci_dss_10.2.4,pci_dss_10.2.5,tsc_CC6.1,tsc_CC6.8,tsc_CC7.2,tsc_CC7.3,</group>
  </rule>

  <rule id="110060" level="12">
    <if_sid>5700</if_sid>
    <match>^Failed|^error: PAM: Authentication</match>
    <srcip>1.1.1.1</srcip>
    <user>superadmin</user>
    <description>sshd: authentication failed from IP $(srcip) as $(srcuser)</description>
    <mitre>
      <id>T1110</id>
    </mitre>
    <group>authentication_failed,gdpr_IV_35.7.d,gdpr_IV_32.2,gpg13_7.1,hipaa_164.312.b,nist_800_53_AU.14,nist_800_53_AC.7,pci_dss_10.2.4,pci_dss_10.2.5,tsc_CC6.1,tsc_CC6.8,tsc_CC7.2,tsc_CC7.3,</group>
  </rule>

</group>
```

I'm using other rule ranges since I have occupied 100001 already.

You can test this rule after you restart the Wazuh Manager service by running a command such as the following:

```
echo 'Nov  8 10:26:22 ip-10-0-1-119 sshd[18780]: Failed password for invalid user superadmin from 1.1.1.1 port 55040 ssh2' | /var/ossec/bin/wazuh-logtest
```

Which gives me the following result:

```
Starting wazuh-logtest v4.3.7
Type one log per line


**Phase 1: Completed pre-decoding.
        full event: 'Nov  8 10:26:22 ip-10-0-1-119 sshd[18780]: Failed password for invalid user superadmin from 1.1.1.1 port 55040 ssh2'
        timestamp: 'Nov  8 10:26:22'
        hostname: 'ip-10-0-1-119'
        program_name: 'sshd'

**Phase 2: Completed decoding.
        name: 'sshd'
        parent: 'sshd'
        srcip: '1.1.1.1'
        srcuser: 'superadmin'

**Phase 3: Completed filtering (rules).
        id: '110060'
        level: '12'
        description: 'sshd: authentication failed from IP 1.1.1.1 as superadmin'
        groups: '['syslog', 'sshd', 'authentication_failed']'
        firedtimes: '1'
        gdpr: '['IV_35.7.d', 'IV_32.2']'
        gpg13: '['7.1']'
        hipaa: '['164.312.b']'
        mail: 'True'
        mitre.id: '['T1110']'
        mitre.tactic: '['Credential Access']'
        mitre.technique: '['Brute Force']'
        nist_800_53: '['AU.14', 'AC.7']'
        pci_dss: '['10.2.4', '10.2.5']'
        tsc: '['CC6.1', 'CC6.8', 'CC7.2', 'CC7.3']'
**Alert to be generated.
```

I hope this helps. Let me know if I'm missing something.

Regards,
Fede

**Reply by f-galland** [reply]
It's worth clarifying that even though wrongly formatted rules might not stop you from saving when you edit them from the Wazuh Dashboard (there are [a few](https://github.com/wazuh/wazuh-kibana-app/issues/4059) known [reported issues](https://github.com/wazuh/wazuh-kibana-app/issues/4422) on the subject)  you would still get a proper warning on `/var/ossec/logs/ossec.log`.
There is sort of a trade off between user experience (the UI shouldn't prevent you from saving an incomplete config file) and proper warning of such an issue.
We expect to improve this situation by release 4.4

**Reply by martinzeifang** [reply]
thank you for your answers. I think you misunderstood my problem. But thanks to that post, I tried to recreate the Problem and it worked 

But I’d like to understand why. 
My problem: 
I have two local xml files. 
First File: (shortened) 
```
<group name="windows-custom,">

  <rule id="100539" level="8">
    <if_sid>100535</if_sid>
    <match>powercat|nc.ps1|nc.exe|mimikatz|tcpdump</match>
    <description>Possible Attacker Software or Command was used</description>
    <group>audit_command,</group>
    <mitre>
        <id>T1003.002</id>
    </mitre>
  </rule>
  
    <rule id="110539" level="15" frequency="3" timeframe="120" ignore="60">
    <if_matched_sid>100539</if_matched_sid>
    <description>Posible Attacker in Network - Multiple Events - Posible Attacker Software or Command was used</description>
    <group>audit_command,</group>
  </rule>

</group>
```

Second File: 
```
<group name="windows-custom,">
 
    <rule id="120539" level="9">
    <if_matched_sid>100539</if_matched_sid>
    <match>powercat</match>
    <description>This is a test for an other xml file</description>
    <group>audit_command,</group>
  </rule>

</group>
```
In the past the rules from the second file never worked. This time i used <if_matched_sid> not <if_sid> can this be the reason why the rule from the second file is working now?

thank you!

**Reply by f-galland** [reply]
Hi Martin,

It's hard to know what's exactly going on without example logs and the full custom rules you are using.
If you are willing to provide those, we can test it on our end.

Regards,
Federico
